src/cmap/background.rs

```diff
@@ -53,13 +75,31 @@ impl ConnectionPool {
     /// each iteration and acquire it again during the next one to ensure that the this method
     /// doesn't block other threads from acquiring connections.
     async fn ensure_min_connections(&self) {
-        if let Some(min_pool_size) = self.inner.min_pool_size {
+        if let Some(min_pool_size) = self.min_pool_size {
             loop {
-                let mut connection_manager = self.inner.connection_manager.lock().await;
-                if connection_manager.total_connection_count < min_pool_size {
-                    match connection_manager.create_connection().await {
-                        Ok(connection) => {
-                            connection_manager.checked_in_connections.push(connection)
+                if self.total_connection_count.load(Ordering::SeqCst) < min_pool_size {
+                    // Reserve a spot via the wait queue. This will prevent too many threads from
+                    // concurrently creating connections such that max_pool_size is exceeded.
+                    let _wait_queue_handle = match self.wait_queue.try_skip_queue() {
+                        None => {
+                            // This branch is rarely taken because we _just_ verified that
+                            // total_connection_count < min_pool_size, which implies
+                            // total_connection_count <= max_pool_size. It is possible however to
+                            // take this branch if a bunch of operation threads enter the queue
+                            // between the count check and our attempt to reserve a spot. In that
+                            // case we just return early because those threads that did enter the
+                            // queue will ensure min_pool_size is fulfilled.
+                            return;
+                        }
+                        Some(handle) => handle,
+                    };
+
+                    let connection = self.create_pending_connection();
+                    match self.establish_connection(connection).await {
+                        Ok(mut connection) => {
+                            let mut available_connections = self.available_connections.lock().await;
```

Comment:
Similar to my comment below about accessing the connection count without holding the lock, I'm starting to wonder if we're opening ourselves up to races in the way that we're using the connection count. We check if we're below minPoolSize above, but we're not holding the lock in the meantime. By the time we actually add the connection to the pool, we've taken the time to not only create the connection but perform the handshake and potentially authenticate as well, so it seems not unlikely that the connection count could have changed between now and then. While it's not necessarily a problem if we populate past minPoolSize, I think we should be checking maxPoolSize while holding the lock before to ensure we don't go over it when adding the connection to the set of available ones.

Reply:
Similar situation here to the other one: we're holding a wait queue handle so we're guaranteed to not exceed maxPoolSize. We definitely could go over minPoolSize, but I didn't really see that as an issue.